packages/truffle-codec/lib/types/pointer.ts

```diff
@@ -0,0 +1,64 @@
+import { AstDefinition } from "truffle-codec-utils";
+import { Range } from "./storage";
+
+export type DataPointer = StackPointer | MemoryPointer | StoragePointer
+  | CalldataPointer | StackLiteralPointer | ConstantDefinitionPointer
+  | SpecialPointer | EventDataPointer | EventTopicPointer;
+
+export type AbiPointer = CalldataPointer | EventDataPointer | GenericAbiPointer;
+export type AbiDataPointer = CalldataPointer | EventDataPointer;
+
+export interface StackPointer {
+  location: "stack",
+  from: number;
+  to: number;
+}
+
+export interface MemoryPointer {
+  location: "memory",
+  start: number;
+  length: number;
+}
+
+export interface CalldataPointer {
+  location: "calldata",
+  start: number;
+  length: number;
+}
+
+export interface EventDataPointer {
+  location: "eventdata",
+  start: number;
+  length: number;
+}
+
+export interface EventTopicPointer {
+  location: "eventtopic",
+  topic: number;
+}
+
+export interface GenericAbiPointer {
+  location: "abi",
+  start: number;
+  length: number;
+}
+
+export interface StoragePointer {
+  location: "storage",
+  range: Range;
+}
+
+export interface StackLiteralPointer {
+  location: "stackliteral",
+  literal: Uint8Array;
+}
+
+export interface ConstantDefinitionPointer {
+  location: "definition",
+  definition: AstDefinition;
+}
+
+export interface SpecialPointer {
+  location: "special",
+  special: string; //sorry
```

Comment:
too much effort to make an enum type of all possible values?

Reply:
Well, so, let's consider -- what are all the possible values?

There's `msg`, `tx`, and `block`, of course.  Those are the ones we currently support.  Then there's `abi`, which we don't support because it just holds functions.

But there's another possibility: `meta`.  Or `meta_type`, or, uh, however we decide it should be interpreted.  Point is, if you write `type(C)`, where `C` is a contract name, you'll get a magic object containing information about that type.  If you do this with a different contract `D`, the result will (according to Solidity) have a different type than if you'd done it with `C`.  So, we probably don't want to consider those as the same type.

We don't currently support metatype, of course, because we have no reason to, but we might want to eventually?  IDK.  If we do, though, an enum won't be good enough, because you'll be able to make a metatype coresponding to any contract.  Well, OK -- we could have it be an enum, and then add an extra field identifying the contract that would only be used in case of `metatype`.  That would work.

But for now I decided to just leave it as a string.  Likely though I should make it an enum, on account of, if we decided to support metatype, a string wouldn't be good enough anyway.  IDK, let me know what you think.

Reply:
Hm, `string` is probably fine for now, then.

Reply:
Huh, really? I was going to go ahead and make it an enum after all, after actually thinking through it above -- like I said, if we really want to handle metatype in the future, a string won't be enough anyway, so I may as well make it an enum.

Reply:
Well, go for it!